File: medium/105. Construct Binary Tree from Preorder and Inorder Traversal.js
// 105. Construct Binary Tree from Preorder and Inorder Traversal

// Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.

// Example 1:


// Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
// Output: [3,9,20,null,null,15,7]
// Example 2:

// Input: preorder = [-1], inorder = [-1]
// Output: [-1]
 

// Constraints:

// 1 <= preorder.length <= 3000
// inorder.length == preorder.length
// -3000 <= preorder[i], inorder[i] <= 3000
// preorder and inorder consist of unique values.
// Each value of inorder also appears in preorder.
// preorder is guaranteed to be the preorder traversal of the tree.
// inorder is guaranteed to be the inorder traversal of the tree.



/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */


// #1
const buildTree = function(preorder, inorder) {
  let preorderIdx = 0
  const inorderIdxMap = {}

  for (let i = 0; i < inorder.length; i++) {
    inorderIdxMap[inorder[i]] = i
  }

  const arrayToTree = (left, right) => {
    if (left > right) {
      return null
    }

    const rootValue = preorder[preorderIdx]
    const root = new TreeNode(rootValue)

    preorderIdx += 1

    root.left = arrayToTree(left, inorderIdxMap[rootValue] - 1)
    root.right = arrayToTree(inorderIdxMap[rootValue] + 1, right)

    return root
  }

  return arrayToTree(0, preorder.length - 1)
};

File: medium/105. Construct Binary Tree from Preorder and Inorder Traversal.py
# 1
class Solution:
  def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    preorderIdx = 0
    inorderIdxMap = {}

    for idx, val in enumerate(inorder):
      inorderIdxMap[val] = idx

    def arrayToTree(left, right):
      nonlocal preorderIdx
      if left > right:
        return None

      rootVal = preorder[preorderIdx]
      root = TreeNode(rootVal)

      preorderIdx += 1

      root.left = arrayToTree(left, inorderIdxMap[rootVal] - 1)
      root.right = arrayToTree(inorderIdxMap[rootVal] + 1, right)

      return root

    return arrayToTree(0, len(preorder) - 1)

# 2
class Solution:
  preorder = []
  preorderIndex = 0
  inorderMap = {}

  def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    self.mapInorder(inorder)
    self.preorder = preorder
    return self.recursiveTree(0, len(inorder))

  def recursiveTree(self, inorderStart, inorderEnd):
    if(inorderEnd <= inorderStart):
      return None
    rootVal = self.preorder[self.preorderIndex]
    root = TreeNode(rootVal)
    rootIndex = self.inorderMap[rootVal]
    self.preorderIndex += 1
    
    if(inorderEnd - inorderStart == 1 ):
      return root
    root.left = self.recursiveTree(inorderStart, rootIndex)
    root.right = self.recursiveTree(rootIndex+1, inorderEnd)
    
    return root
    
  def mapInorder(self, inorder):
    for i in range(len(inorder)):
      self.inorderMap[inorder[i]] = i 